### PR TITLE
Extract message serialization from `TcpTransport`

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit introduces a `NetworkMessage` class. This class has two
subclasses - `InboundMessage` and `OutboundMessage`. These messages can
be serialized and deserialized independent of the transport. This allows
more granular testing. Additionally, the serialization mechanism is now
a simple `Supplier`. This builds the framework to eventually move the
serialization of transport messages to the network thread. This is the
one serialization component that is not currently performed on the
network thread (transport deserialization and http serialization and
deserialization are all on the network thread).

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run gradle build tests 2

File: server/src/main/java/org/elasticsearch/transport/InboundMessage.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        // once the connection is established
+        final Version compatibilityVersion = isHandshake ? currentVersion.minimumCompatibilityVersion() : currentVersion;
+        if (version.isCompatible(compatibilityVersion) == false) {
+            final Version minCompatibilityVersion = isHandshake ? compatibilityVersion : compatibilityVersion.minimumCompatibilityVersion();
+            String msg = "Received " + (isHandshake ? "handshake " : "") + "message from unsupported version: [";
+            throw new IllegalStateException(msg + version + "] minimal compatible version is: [" + minCompatibilityVersion + "]");
+        }
+    }
+
+    public static class Request extends InboundMessage {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/InboundMessage.java]: I'd name these `ReqeustMessage` and `ResponseMessage`

File: server/src/main/java/org/elasticsearch/transport/NetworkMessage.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.transport;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+
+public abstract class NetworkMessage {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/NetworkMessage.java]: a class level javadoc would be great.

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+import org.elasticsearch.common.metrics.MeanMetric;
+import org.elasticsearch.common.network.CloseableChannel;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.core.internal.io.IOUtils;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+class OutboundHandler {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/OutboundHandler.java]: can we final?

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+        private SendContext(TcpChannel channel, CheckedSupplier<BytesReference, IOException> messageSupplier,
+                            ActionListener<Void> listener, Releasable optionalReleasable) {
+            this.channel = channel;
+            this.messageSupplier = messageSupplier;
+            this.listener = listener;
+            this.optionalReleasable = optionalReleasable;
+        }
+
+        public BytesReference get() throws IOException {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/OutboundHandler.java]: can be pkg private

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            BytesReference message;
+            try {
+                message = messageSupplier.get();
+                messageSize = message.length();
+                transportLogger.logOutboundMessage(channel, message);
+                return message;
+            } catch (IOException | RuntimeException e) {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/OutboundHandler.java]: you can just catch `Exception` here and rethrow.

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+        @Override
+        protected void innerOnFailure(Exception e) {
+            logger.warn(() -> new ParameterizedMessage("send message failed [channel: {}]", channel), e);
+            closeAndCallback(e, () -> listener.onFailure(e));
+        }
+
+        private void closeAndCallback(final Exception e, Runnable runnable) {
+            try {
+                IOUtils.close(optionalReleasable, runnable::run);
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/transport/OutboundHandler.java]: I'd just use `closeWhileHandlingException` instead.

**Review by s1monw — APPROVED** [review]
I did a rather high level review. Sorry for taking so long. I think it's good to make TcpTransport less complicated and add these dedicated tests, good! Most of it is just moving code around, I am glad you are working on it.

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java (hunk 4 above)
Comment (tbrooks8) [reply server/src/main/java/org/elasticsearch/transport/OutboundHandler.java]: No I don't think so. This is implementing the `CheckedSupplier` interface.

**Pull request merged**